pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java

```diff
@@ -0,0 +1,196 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.io.util;
+
+import com.google.common.base.Preconditions;
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Arrays;
+import org.apache.pinot.common.utils.StringUtil;
+import org.apache.pinot.core.segment.memory.PinotDataBuffer;
+
+/**
+ * Implementation of {@link ValueReader} that will allow each value to be of variable
+ * length and there by avoiding the unnecessary padding.
+ *
+ * The layout of the file is as follows:
+ * <p> Header Section: </p>
+ * <ul>
+ *   <li> Magic header byte sequence. </li>
+ * </ul>
+ *
+ * <p> Values: </p>
+ * <ul>
+ *   <li> Integer offsets to start position of byte arrays. </li>
+ *   <li> All byte arrays. </li>
+ * </ul>
+ *
+ * Only sequential writes are supported.
+ *
+ * @see FixedByteValueReaderWriter
+ */
+public class VarLengthBytesValueReaderWriter implements Closeable, ValueReader {
```

Comment:
The convention is to use ReaderWriter for mutable segments, and those classes implement both Reader and writer (IIRC). This naming is breaking that convention and would create confusion.

Reply:
Yes. I've discuss this with @kishoreg in-person but the gist of it is, this new class is named exactly on the lines of an existing class `FixedByteValueReaderWriter` to be consistent. I'll do a new PR cleaning up both of these.

Reply:
Please file an issue for all cleanups not being addressed in this PR, so we can track them.

Reply:
Filed https://github.com/apache/incubator-pinot/issues/4393 to do the cleanup.